mantidproject/mantid
# Perform reduction on refl GUI when ROI changed

This PR performs a full reduction in the ISIS reflectometry GUI Preview tab. The reduction is performed when the region selection has changed.

The main changes are:

- The `PreviewPresenter` gets the selected ROI from the view and sets it on the `PreviewModel`. It then tells the model to start the reduction.
- The model calls the `PreviewJobManager` which uses the `ReflAlgorithmFactory` to create the algorithm, which the job manager then executes asynchronously.
- The `algorithmComplete` notification on the `PreviewJobManager` notifies the presenter when the reduction is finished. For now this just prints a message to the log. 

**To test:**

This must be run in a development build.

Note that this is work in progress on a hidden GUI component so is very rough and ready at the moment. The look and usability of these features is currently very poor so you do not need to test that side thoroughly; just ensure it performs as expected in the happy path case and does not crash etc.

- Ensure this workspace is loaded or is in your load path: [INTER45455_inst.zip](https://github.com/mantidproject/mantid/files/7504266/INTER45455_inst.zip)
- Open the ISIS Reflectometry GUI
- Select the Preview tab.
- Type in `INTER45455_inst`, set the angle to `0.5`, and click Load. The instrument view plot should display the data. Click the rectangle-select button above it and select a region. The second plot will be updated.
- Click the rectangle-selection on the second plot and select a region. A reduction will now be performed on the selected spectra and a log message `Reduction completed` should be displayed.
  - Note that the reduction is quite picky and may fail; if so you will get a warning in the log. It's best to select a region around spectra 20-30.

![image](https://user-images.githubusercontent.com/12895056/177745550-d54aca7d-f0f3-4630-a15d-658a58813890.png)

Part of #34106 

*This does not require release notes* because **it only affects functionality that is not yet available in the release build**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
